Dart/src/com/jetbrains/lang/dart/ide/hierarchy/call/DartCalleeTreeStructure.java

```diff
@@ -0,0 +1,165 @@
+                throw new ExitVisitor();
+              }
+              super.visitMethodDeclaration(element);
+            }
+
+            public void visitFunctionDeclarationWithBodyOrNative(@NotNull DartFunctionDeclarationWithBodyOrNative element) {
+              if (element.getTextOffset() == offset) {
+                result[0] = element;
+                throw new ExitVisitor();
+              }
+              super.visitFunctionDeclarationWithBodyOrNative(element);
+            }
+
+            public void visitFunctionDeclarationWithBody(@NotNull DartFunctionDeclarationWithBody element) {
+              if (element.getTextOffset() == offset) {
+                result[0] = element;
+                throw new ExitVisitor();
+              }
+              super.visitFunctionDeclarationWithBody(element);
+            }
+
+            public void visitSetterDeclaration(@NotNull DartSetterDeclaration element) {
+              if (element.getTextOffset() == offset) {
+                result[0] = element;
+                throw new ExitVisitor();
+              }
+              super.visitSetterDeclaration(element);
+            }
+
+            public void visitGetterDeclaration(@NotNull DartGetterDeclaration element) {
+              if (element.getTextOffset() == offset) {
+                result[0] = element;
+                throw new ExitVisitor();
+              }
+              super.visitGetterDeclaration(element);
+            }
+
+            public void visitFactoryConstructorDeclaration(@NotNull DartFactoryConstructorDeclaration element) {
+              if (element.getTextOffset() == offset) {
+                result[0] = element;
+                throw new ExitVisitor();
+              }
+              super.visitFactoryConstructorDeclaration(element);
+            }
+
+            public void visitNamedConstructorDeclaration(@NotNull DartNamedConstructorDeclaration element) {
+              if (element.getTextOffset() == offset) {
+                result[0] = element;
+                throw new ExitVisitor();
+              }
+              super.visitNamedConstructorDeclaration(element);
+            }
+          });
+        }
+        catch (ExitVisitor ex) {
+          return result[0];
+        }
+      }
+    }
+    return null;
+  }
+
+  @NotNull
+  private static List<DartNavigationTarget> getRegionAt(int offset, @NotNull List<DartNavigationRegion> regions) {
+    for (DartNavigationRegion region : regions) {
+      int targetStart = region.getOffset();
+      int targetEnd = targetStart + region.getLength();
+      if (offset >= targetStart && offset <= targetEnd) {
+        return region.getTargets();
+      }
+    }
+    return new ArrayList<>(0);
+  }
+
+  @NotNull
+  @Override
+  protected List<PsiElement> getChildren(@NotNull PsiElement element) {
+    final List<PsiElement> list = new ArrayList<>();
+    getCallees(element, list);
+    return list;
+  }
+
+  private static class ExitVisitor extends Error {
```

Comment:
I suppose I could have re-used `org.apache.velocity.runtime.directive.StopCommand`, but adding that dependency seemed wrong. This is such a common pattern (outside the Jetbrains codebase ;-) that it seemed more intuitive to define a new class.